File: Recorder/src/main/java/com/github/ezauton/recorder/base/RobotStateRecorder.java
package com.github.ezauton.recorder.base;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.github.ezauton.core.localization.RotationalLocationEstimator;
import com.github.ezauton.core.localization.TranslationalLocationEstimator;
import com.github.ezauton.core.utils.Clock;
import com.github.ezauton.recorder.SequentialDataRecorder;
import com.github.ezauton.recorder.base.frame.RobotStateFrame;

import java.util.concurrent.TimeUnit;

public class RobotStateRecorder extends SequentialDataRecorder<RobotStateFrame> {

    private static int instanceCounter = 0;
    @JsonIgnore
    private TranslationalLocationEstimator posEstimator;
    @JsonIgnore
    private RotationalLocationEstimator rotEstimator;
    private double width;
    private double height;

    public RobotStateRecorder(String name, Clock clock, TranslationalLocationEstimator posEstimator, RotationalLocationEstimator rotEstimator, double width, double length) {
        super(name, clock);
        this.posEstimator = posEstimator;
        this.rotEstimator = rotEstimator;
        this.width = width;
        this.height = length;
    }

    public RobotStateRecorder(Clock clock, TranslationalLocationEstimator posEstimator, RotationalLocationEstimator rotEstimator, double width, double length) {
        this("RobotStateRecorder_" + instanceCounter++, clock, posEstimator, rotEstimator, width, length);
    }

    private RobotStateRecorder() {
        super();
    }

    @Override
    public boolean checkForNewData() {
        dataFrames.add(new RobotStateFrame(
                stopwatch.read(TimeUnit.MILLISECONDS),
                posEstimator.estimateLocation(),
                rotEstimator.estimateHeading(),
                width,
                height,
                posEstimator.estimateAbsoluteVelocity()
        ));
        return true;
    }

//    @Override
//    public IDataProcessor createDataProcessor()
//    {
//        return new RobotStateDataProcessor(this);
//    }
}

File: Recorder/src/main/java/com/github/ezauton/recorder/base/RobotStateRecorder.kt
package com.github.ezauton.recorder.base

//import com.github.ezauton.recorder.SubRecording
//import com.github.ezauton.recorder.base.frame.RobotStateFrame
//import kotlinx.serialization.SerialName
//import kotlinx.serialization.Serializable
//
//
//@Serializable
//@SerialName("robotState")
//class RobotStateRecording(val frames: List<RobotStateFrame>, override val name: String) : SubRecording
//
//
//fun robotStateRecorder(posEstimator: TranslationalLocationEstimator, rotEstimator: RotationalLocationEstimator, width: Double, height: Double, period: PeriodicParams = PeriodicParams.DEFAULT) =
//
//
//  sendAction {
//    val frames =  periodic(period) { loop ->
//      RobotStateFrame(
//        posEstimator.estimateLocation().scalarVector,
//        rotEstimator.estimateHeading().value,
//        width,
//        height,
//        posEstimator.estimateAbsoluteVelocity().scalarVector,
//        loop.stopwatch.read().millis,
//      )
//    }
//
//    println("caught with!!!")
//
//    emit(RobotStateRecording(frames, "RobotStateRecording"))
//
//  }
